src/platformdirs/windows.py

```diff
@@ -161,108 +167,165 @@ def get_win_folder_from_env_vars(csidl_name: str) -> str:
+        """
+        `
+        The GUID structure from Windows's guiddef.h header
+        <https://learn.microsoft.com/en-us/windows/win32/api/guiddef/ns-guiddef-guid>`_.
+        """
 
-    return buf.value
+        Data4Type = ctypes.c_ubyte * 8
+
+        _fields_ = (
+            ("Data1", ctypes.c_ulong),
+            ("Data2", ctypes.c_ushort),
+            ("Data3", ctypes.c_ushort),
+            ("Data4", Data4Type),
+        )
+
+        def __init__(self, guid_string: str) -> None:
+            digit_groups = guid_string.split("-")
+            expected_digit_groups = 5
+            if len(digit_groups) != expected_digit_groups:
+                msg = f"The guid_string, {guid_string!r}, does not contain {expected_digit_groups} groups of digits."
+                raise ValueError(msg)
+            for digit_group, expected_length in zip(digit_groups, (8, 4, 4, 4, 12)):
+                if len(digit_group) != expected_length:
+                    msg = (
+                        f"The digit group, {digit_group!r}, in the guid_string, {guid_string!r}, was the wrong length. "
+                        f"It should have been {expected_length} digits long."
+                    )
+                    raise ValueError(msg)
+            data_4_as_bytes = bytes.fromhex(digit_groups[3]) + bytes.fromhex(digit_groups[4])
+
+            super().__init__(
+                int(digit_groups[0], base=16),
+                int(digit_groups[1], base=16),
+                int(digit_groups[2], base=16),
+                self.Data4Type(*(eight_bit_int for eight_bit_int in data_4_as_bytes)),
+            )
+
+        def __repr__(self) -> str:
+            guid_string = f"{self.Data1:08X}-{self.Data2:04X}-{self.Data3:04X}-"
+            for i in range(len(self.Data4)):
+                guid_string += f"{self.Data4[i]:02X}"
+                if i == 1:
+                    guid_string += "-"
+            return f"{type(self).__qualname__}({guid_string!r})"
+
+    def get_win_folder_via_ctypes(folderid_name: str) -> str:
+        """Get folder with ctypes."""
+        # https://learn.microsoft.com/en-us/windows/win32/shell/knownfolderid
+        folderid_const = {
+            "FOLDERID_RoamingAppData": GUID("3EB685DB-65F9-4CF6-A03A-E3EF65729F3D"),
+            "FOLDERID_ProgramData": GUID("62AB5D82-FDC1-4DC3-A9DD-070D1D495D97"),
+            "FOLDERID_LocalAppData": GUID("F1B32785-6FBA-4FCF-9D55-7B8E7F157091"),
+            "FOLDERID_Documents": GUID("FDD39AD0-238F-46AF-ADB4-6C85480369C7"),
+            "FOLDERID_Pictures": GUID("33E28130-4E1E-4676-835A-98395C3BC3BB"),
+            "FOLDERID_Videos": GUID("18989B1D-99B5-455B-841C-AB7C74E4DDFC"),
+            "FOLDERID_Music": GUID("4BD8D571-6D19-48D3-BE97-422220080E43"),
+            "FOLDERID_Downloads": GUID(FOLDERID_Downloads_guid_string),
+            "FOLDERID_Desktop": GUID("B4BFCC3A-DB2C-424C-B029-7FE99A87C641"),
+        }.get(folderid_name)
+        if folderid_const is None:
```

Comment:
This is doing a lot of work now, please do `if-elif-else` instead.

Reply:
OK, done.